File: android/src/main/java/com/iconshot/detonator/emitter/HandlerEmitter.java
package com.iconshot.detonator.emitter;

import com.iconshot.detonator.Detonator;

public class HandlerEmitter extends Emitter {
    public HandlerEmitter(Detonator detonator) {
        super(detonator);
    }

    public void emit(String name, int edgeId, Object data) {
        Handler handler = new Handler();

        handler.name = name;
        handler.edgeId = edgeId;
        handler.data = data;

        detonator.emit("handler", handler);
    }

    private static class Handler {
        String name;
        int edgeId;
        Object data;
    }
}

File: ios/Sources/Detonator/Emitter/HandlerEmitter.swift
public class HandlerEmitter: Emitter {
    public func emit(name: String, edgeId: Int, data: Encodable?) {
        let handler = Handler(name: name, edgeId: edgeId, data: data)
        
        detonator.emit(name: "handler", value: handler)
    }
    
    struct Handler: Encodable {
        let name: String
        let edgeId: Int
        let data: AnyEncodable??

        init(name: String, edgeId: Int, data: Encodable?) {
            self.name = name
            self.edgeId = edgeId
            
            if let data = data {
                self.data = AnyEncodable(data)
            } else {
                self.data = .some(nil)
            }
        }
    }
}
